lib/recursive-methods.rb

```diff
@@ -1,49 +1,108 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(n)
+# source: video example in class
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0
+        raise ArgumentError
+    elsif n == 1 || n == 0
+        return 1 
+    else
+        return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s 
+    else
+        return s[-1] + reverse(s[1...-1]) + s[0]
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: 0(n)
+# Space complexity: 0(1)
+def reverse_inplace(s, low = 0, high = s.length - 1) 
```

Comment:
Your space complexity will be O(n) because of the call stack, but otherwise correct.